doc/BorderRouter.md

```diff
@@ -0,0 +1,124 @@
+The communication between those layers are implemented as go channels.
+
+* **Receivers** There is one receiver per border router interface that is responsible for batch-reading the
+packets from the network socket, identifying the source and flowID and using them to identify which processing
+routine has to process the packet.
+Then the receiver enqueues the packet to that processing routine.
+If the queue of that processing routine is full, the packet will be dropped.
+Each receiver has a pool of preallocated packet buffers that they can use to store the packets they receive.
+This can be implemented as a go channel of packets where the receiver reads a certain amount of packets,
+updates the pointers of the ipv4.Message.Buffers to the buffer inside those packets and then performs a batch read.
+* **Processing Routines** There are several processing routines and slow-path processing routines
+in the border router that are responsible for processing the received packet.
+The actual processing logic remains unchanged.
+If the processing routine identifies a packet that belongs to the slow-path, the processing routines enqueues
+the packet to a slow-path processing routine. If the queue of the slow-path processing routine is full, the
+packet will not be processed at all. In this case the buffer is immediately returned to the receiver.
+Once a packet is processed, it gets enqueued to the forwarder which is responsible for the egress interface.
+If the queue of the forwarder is full, the packet will be dropped.
+* **Forwarders** There exists one forwarder per border router interface that is responsible for forwarding the
+packets over the network that it receives from the processing routines. It forwards the packets as a batch.
+Afterwards it returns each buffer to the receiver from which it originates.
+
+![Border Router Design](fig/border_router/br_design.png)
+
+## Mapping of processing routines
+To prevent any packet reordering on the fast-path, we map the tuple of source and flowID, see the
+[SCION header documentation](https://github.com/scionproto/scion/blob/master/doc/protocols/scion-header.rst),
+to a fixed processing routine using a hash function together with a secret value to prevent pre-computations
+of the exact mapping.
+
+## Slow path
+During processing, packets that have to follow the slow path are identified and forwarded to the
+slow-path processing routines.
+Rate limiting of slow-path operations is not implemented explicitly, but only implictily through specifying
+the number of slow-path processing routines in the configuration.
+In case a packet is identified to belong to the slow path but the queue of the slow path is full, the
+packet is dropped.
+Packets currently identified for slow-path are:
+* Malformed packets
+* SCMP traceroute packets
+
+# Configuration
+The configuration of the border router will remain in the border router toml file.
+The following configuration entries are added:
+
+## Pool size
```

Comment:
I find this parameter okay _initially_. But add a note saying that this value must be derivable from the max latency that the processing routine + forwarder need to process a packet, and the speed of the interface.

Reply:
Maybe write that it depends on the speed of the network **interfaces**?
The pool size not only depends on the speed of the interface where the receiver gets packets from, it also depends on the speed of the interfaces to which the packets are sent.
If they are very slow, then we do not need a large pool, even if packets arrive at a high rate -> the packets would just be dropped when trying to enqueue them and be available in the pool again.
If this argument makes sense to you, you might also want to explain this in the document

Reply:
Yes, that's what I mean with 
> derivable from the max latency that the processing routine + forwarder need to process a packet, and the speed of the interface

More specifically, the pool size is a function such as
```
pool_size = f(receiver_interface, processing_routine_max_lat, forwarder_max_lat){...}
```
Where all three parameters can be estimated by the BR itself, on startup, or using a specific switch.

But again, I'd be okay with just a note for now, and setting the value specifically in the configuration.